framework/decode/file_processor.cpp

```diff
@@ -494,27 +494,46 @@ bool FileProcessor::ProcessBlocks()
 // the correct sizing of the block payload are done by the caller
 bool FileProcessor::ReadBlockBuffer(BlockParser& parser, BlockBuffer& block_buffer)
 {
-    bool success = parser.ReadBlockBuffer(GetCurrentFile().active_file, block_buffer);
-    if (success)
+    bool           success = true;
+    BlockReadError status  = parser.ReadBlockBuffer(GetCurrentFile().active_file, block_buffer);
+    if (status == kErrorNone)
     {
         bytes_read_ += block_buffer.Size();
     }
     else
     {
-        HandleBlockReadError(kErrorReadingBlockData, "Failed to read block body data");
+        // Caller handles end of file on block boundaries
+        if (status != kEndOfFile)
+        {
+            HandleBlockReadError(status, "Failed to read next block");
+        }
+        success = false;
     }
     return success;
 }
 
 // Parse the block header and load the whole block into a block buffer
-bool BlockParser::ReadBlockBuffer(FileInputStreamPtr& input_stream, BlockBuffer& block_buffer)
+BlockReadError BlockParser::ReadBlockBuffer(FileInputStreamPtr& input_stream, BlockBuffer& block_buffer)
 {
     using BlockSizeType = decltype(format::BlockHeader::size);
     BlockSizeType block_size;
-    bool          success = input_stream->PeekBytes(&block_size, sizeof(block_size));
-    if (success)
+    BlockReadError status = kErrorNone;
+
+    if (!input_stream->PeekBytes(&block_size, 1))
```

Comment:
Aren't you copying a byte here on this check if it succeeds.  So aren't you doing extra work each call?  Why not just incorporate the check in PeekBytes?